edx_notifications/base_data.py

```diff
@@ -0,0 +1,287 @@
+class DictField(TypedField):
+    """
+    Specialized subclass of TypedField(dict) as a convienence
+    """
+
+    def __init__(self, **kwargs):
+        super(DictField, self).__init__(dict, kwargs)
+
+    @classmethod
+    def to_json(cls, data):
+        """
+        Serialize to json
+        """
+
+        if not data:
+            return None
+
+        _dict = copy.deepcopy(data)
+
+        for key, value in _dict.iteritems():
+            if isinstance(value, datetime):
+                _dict[key] = value.isoformat()
+
+        return json.dumps(_dict)
+
+    @classmethod
+    def from_json(cls, value):
+        """
+        Descerialize from json
+        """
+
+        if not value:
+            return None
+
+        _dict = json.loads(value)
+
+        for key, value in _dict.iteritems():
+            if isinstance(value, basestring):
+                # This could be a datetime posing as a ISO8601 formatted string
+                # we so have to apply some heuristics here
+                # to see if we want to even attempt
+                if value.count('-') == 2 and value.count(':') == 2 and value.count('T') == 1:
+                    # this is likely a ISO8601 serialized string, so let's try to parse
+                    try:
+                        _dict[key] = dateutil.parser.parse(value)
+                    except ValueError:
+                        # oops, I guess our heuristic was a bit off
+                        # no harm, but just wasted CPU cycles
+                        pass
+
+        return _dict
+
+
+class DateTimeField(TypedField):
+    """
+    Specialized subclass of TypedField(datetime) as a convienence
+    """
+
+    def __init__(self, **kwargs):
+        super(DateTimeField, self).__init__(datetime, kwargs)
+
+
+class EnumField(StringField):
+    """
+    Specialized subclass of TypedField() which is basically an StringTypedField,
+    but constrains what values can be set on it
+    """
+
+    def __init__(self, **kwargs):
+        """
+        Initializer with constrained values
+        """
+
+        self._allowed_values = kwargs['allowed_values']
+        super(EnumField, self).__init__(**kwargs)
+
+    def __set__(self, instance, value):
+        """
+        Descriptor setter. Be sure to enforce type on setting. But None is allowed.
+        """
+
+        if self._allowed_values:
+            if value not in self._allowed_values:
+                msg = (
+                    "Attempting to set to '{value}'. Allowed values are: '{allowed}'."
+                ).format(value=value, allowed=str(self._allowed_values))
+                raise ValueError(msg)
+
+        super(EnumField, self).__set__(instance, value)
+
+
+class BaseDataObject(object):
+    """
+    A base class for all Notification Data Objects
+    """
+
+    id = IntegerField(default=None)  # pylint: disable=invalid-name
+
+    _is_loaded = False
+    _is_dirty = False  # Returns if this object has been modified after initialization
```

Comment:
`_is_dirty` is never used. Also, from experience, detecting `dirty` when you have mutable fields is a pain.